### PR TITLE
Add env. var. for efficient text-generation in inference

### PR DESCRIPTION
**Pull request by nzmora-nvidia**
In ONNX inference with KV-Cache optimizations for GPT text generation, the causal attention mask can be square (context-phase) or rectangular (generation-phase).
This commit forces softmax to use forward_torch_softmax and to dynamically slice a pre-allocated attention mask - when exporting to ONNX.

* Introduce an environment variable (NVTE_ONNX_KVCACHE_MAX_SEQ_LEN) to set the maximum sequence length.
When exporting to ONNX and this variable is set, TE preallocates an upper triangular (k=1) matrix with a size as prescribed by the variable, and dynamically slices the mask for the required shape.
TE models can be exported to ONNX when NVTE_ONNX_KVCACHE_MAX_SEQ_LEN is not configured, but the attention masking is always square and not fit for efficient text generation.

* Work-around torch.onnx.export bug that incorrectly folds layer_norm(data, scale=add(gamma,1)) to layer_norm(data, scale=gamma) when we use LN with zero-centered gamma.

* ONNX export tests improvements
  * Add a fixture (seed_default_rng) to seed the PRNG
  * Add a fixture (set_max_seq_len) to set the max sequence length when exporting to ONNX for GPT text generation
  * Add dynamic shapes for ONNX input/output tests.
  * Allow validate_result to compare ORT output to pre-computed TE outputs.
  * Add to test_export_softmax a test for te.softmax.FusedScaleMaskSoftmax.
  * Add test_softmax_mask_fn to test that TE's default attention mask and the new ONNX-compatible mask produce the same behavior.
  * Add test_export_gpt_generation to test that the ONNX model can correctly handle inputs with different shapes and that the attention mask it adjusted on-the-fly to different sequence lengths.

### CODE REVIEW

File: tests/pytorch/test_onnx_export.py [hunk 1]
```diff
@@ -81,8 +93,9 @@ def do_export(
     fname: str,
     use_fp8: bool=True,
     opset: int=OPSET,
-    input_names: list=["input"],
-    output_names: list=["output"],
+    input_names: List=["input"],
+    output_names: List=["output"],
```
Comment (ksivaman) [inline tests/pytorch/test_onnx_export.py:97]: Could we change these to `None` to have immutable default values just a good practice? This is for throughout the file, just making one comment.

Comment (nzmora-nvidia) [reply tests/pytorch/test_onnx_export.py:97]: Of course - thanks for point this out!

File: tests/pytorch/test_onnx_export.py [hunk 2]
```diff
@@ -71,6 +71,18 @@
 skip_FP8 = pytest.mark.skipif(not fp8_available, reason=reason_for_no_fp8)
 
 
+@pytest.fixture()
+def seed_default_rng():
+    """Reseed the PRNG for test repeatability"""
```
Comment (ksivaman) [inline tests/pytorch/test_onnx_export.py:76]:
```suggestion
    """Reseed the PRNG for test reproducibility"""
```

**Review by ksivaman — CHANGES_REQUESTED** [review]

File: tests/pytorch/test_onnx_export.py (hunk 2 above)
Comment (nzmora-nvidia) [reply tests/pytorch/test_onnx_export.py]: done

File: transformer_engine/pytorch/softmax.py [hunk 3]
```diff
@@ -25,6 +25,24 @@ def _get_default_causal_mask(sq: int) -> torch.Tensor:
     return _default_causal_mask[sq]
 
 
+def _get_onnx_export_causal_mask(seq_q: int, seq_k: int, onnx_causal_mask) -> torch.Tensor:
```
Comment (ksivaman) [inline transformer_engine/pytorch/softmax.py:28]:
```suggestion
def _get_onnx_export_causal_mask(seq_q: int, seq_k: int, onnx_causal_mask: torch.Tensor) -> torch.Tensor:
```

**Comment by ksivaman** [comment]
Pipeline: 8272095

**Comment by nzmora-nvidia** [comment]
@ksivaman I think all comments have been addressed

**Review by ksivaman — APPROVED** [review]
